### PR TITLE
TagHelper API Cleanup

### PR DESCRIPTION
**Pull request by NTaylorMullen**
- Changed `Microsoft.AspNet.Razor.TagHelpers` to `Microsoft.AspNet.Razor.Compilation.TagHelpers`.
- Changed `Microsoft.AspNet.Razor.Runtime.TagHelpers` to `Microsoft.AspNet.Razor.TagHelpers`.
- Updated folder structure to reflect new namespaces.
- Updated generated code files to reflect new runtime type namespaces.

#578

### CODE REVIEW

**Comment by NTaylorMullen** [comment]
/cc @Eilon @rynowak @dougbu @DamianEdwards 

If all looks good i'll do the reaction to MVC/RazorTooling etc.

**Comment by Eilon** [comment]
Looks :shipit: to me, but would like at least one of @dougbu and @rynowak to confirm this matches their expectations.

**Comment by dougbu** [comment]
Looks like doc comments need to change:
```
Invalid documentation syntax in src\Microsoft.AspNet.Razor\bin\Debug\dnx451\Microsoft.AspNet.Razor.xml
```

File: src/Microsoft.AspNet.Razor/Chunks/AddTagHelperChunk.cs [hunk 1]
```diff
@@ -4,12 +4,12 @@
 namespace Microsoft.AspNet.Razor.Chunks
 {
     /// <summary>
-    /// A <see cref="Chunk"/> used to look up <see cref="TagHelpers.TagHelperDescriptor"/>s.
+    /// A <see cref="Chunk"/> used to look up <see cref="Compilation.TagHelperDescriptor"/>s.
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Chunks/AddTagHelperChunk.cs]: `Compilation.TagHelpers.TagHelperDescriptor`

**Comment by dougbu** [comment]
Why are a few classes e.g. `RuntimePropertyInfo` and `XMLDocumentationProvider` (hmm, shouldn't that be `XmlDocumentationProvider`?) still in the `Microsoft.AspNet.Razor.Runtime` namespace? Should be the case that nothing except resources are left there.

**Comment by dougbu** [comment]
Suggest we follow up by moving infrastructure classes that tag helper authors don't need into `...TagHelpers.Internal`. This would include e.g. `DefaultTagHelperContent`, `TagHelperRunner` and `TagHelperDescriptorResolver`. Tag helper authors should only need the attributes, `TagHelper`, and what they reference.

**Comment by dougbu** [comment]
;watch:

**Comment by NTaylorMullen** [comment]
@dougbu The `RuntimePropertyInfo` type classes aren't `TagHelper` specific. Thoughts on moving them to `Microsoft.AspNet.Razor.Runtime.Internal`?

**Comment by dougbu** [comment]
> The `RuntimePropertyInfo` type classes aren't `TagHelper` specific.

I was looking at an inconsistency. Some of these `...Info` classes moved and a few did not.

In any case, while these classes have no inherent ties to `TagHelper`s, aren't they used only in that context?

> moving them to `Microsoft.AspNet.Razor.Runtime.Internal`?

I suggested that as a follow-up. It's a fine idea now as well but I'd prefer to cover everything that `TagHelper` authors never need (e.g. the factories) in one pass.

**Comment by NTaylorMullen** [comment]
Updated.

File: src/Microsoft.AspNet.Razor/CodeGenerators/Visitors/CSharpUsingVisitor.cs [hunk 2]
```diff
@@ -93,11 +97,15 @@ protected override void Visit(TagHelperChunk chunk)
             {
                 _foundTagHelpers = true;
 
-                if (!ImportedUsings.Contains(TagHelpersRuntimeNamespace))
+                foreach (var tagHelperRuntimeNamespace in TagHelpersRuntimeNamespaces)
                 {
-                    // If we find TagHelpers then we need to add the TagHelper runtime namespace to our list of usings.
-                    Writer.WriteUsing(TagHelpersRuntimeNamespace);
-                    ImportedUsings.Add(TagHelpersRuntimeNamespace);
+                    if (!ImportedUsings.Contains(tagHelperRuntimeNamespace))
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/CodeGenerators/Visitors/CSharpUsingVisitor.cs]: Drive-by suggestion but could `ImportedUsings` be a `HashSet<string>`? This would then simplify to
``` c#
if (ImportedUsings.Add(tagHelperRuntimeNamespace))
{
    Writer.WriteUsing(tagHelperRuntimeNamespace);
}
```

File: test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/DefaultTagHelperContentTest.cs [hunk 3]
```diff
@@ -1,14 +1,13 @@
...
-using System;
 using System.Globalization;
 using System.IO;
-using Microsoft.AspNet.Testing;
+using Microsoft.AspNet.Razor.TagHelpers;
 using Microsoft.Extensions.WebEncoders.Testing;
 using Xunit;
 
-namespace Microsoft.AspNet.Razor.Runtime.TagHelpers
+namespace Microsoft.AspNet.Razor.Runtime.TagHelpers.Internal
```
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/DefaultTagHelperContentTest.cs]: Looks as if test classes are in the correct namespaces but not the correct directories.

File: src/Microsoft.AspNet.Razor/CodeGenerators/Visitors/CSharpUsingVisitor.cs [hunk 4]
```diff
@@ -11,7 +11,11 @@ namespace Microsoft.AspNet.Razor.CodeGenerators.Visitors
 {
     public class CSharpUsingVisitor : CodeVisitor<CSharpCodeWriter>
     {
-        private const string TagHelpersRuntimeNamespace = "Microsoft.AspNet.Razor.Runtime.TagHelpers";
+        private static readonly string[] TagHelpersRuntimeNamespaces = new[]
+        {
+            "Microsoft.AspNet.Razor.TagHelpers",
+            "Microsoft.AspNet.Razor.Runtime.TagHelpers.Internal"
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/CodeGenerators/Visitors/CSharpUsingVisitor.cs]: Why does the generated code need the `Internal` namespace? Does this make the types show up in IntelliSense when the user is in C# mode? I expected more of the explicit type qualification used elsewhere in the generated code.

Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/CodeGenerators/Visitors/CSharpUsingVisitor.cs]: It's identical to what we had before. We need the `Internal` namespace to enable access to the variety of `TagHelper` enabling types such as `TagHelperExecutionContext`, `TagHelperRunner` etc. We could at some point remove these namespaces and generate all the required fields using their fully qualified names.

File: test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/DefaultTagHelperContentTest.cs (hunk 3 above)
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/DefaultTagHelperContentTest.cs]: Done, for some reason git diff  doesn't show unless you look at all all file changes.

**Comment by NTaylorMullen** [comment]
Updated.

File: src/Microsoft.AspNet.Razor/CodeGenerators/Visitors/CSharpUsingVisitor.cs [hunk 5]
```diff
@@ -89,15 +93,23 @@ protected override void Visit(UsingChunk chunk)
 
         protected override void Visit(TagHelperChunk chunk)
         {
+            if (Context.Host.DesignTimeMode)
+            {
+                return;
+            }
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/CodeGenerators/Visitors/CSharpUsingVisitor.cs]: :clap:

**Comment by dougbu** [comment]
Yah know, well I just thought, and I realize you're going to absolutely love this... Could now use `Microsoft.AspNet.Razor.Runtime.TagHelpers` where you've got `Microsoft.AspNet.Razor.Runtime.TagHelpers.Internal`. Ignore me if you're thinking of killing me.

**Comment by dougbu** [comment]
:shipit: after fixing up an MVC fix-up PR.
